core/src/main/kotlin/chat/rocket/core/internal/rest/RestResult.kt

```
@@ -0,0 +1,87 @@
+import java.lang.reflect.Type
+
+class RestResult<out T>(val result: T?, val total: Long?, val offset: Long?, val count: Long?) {
+
+    class MoshiJsonAdapter<T> internal constructor(moshi: Moshi, types: Array<Type>) : JsonAdapter<RestResult<T>>() {
+        private val adapter: JsonAdapter<T>
+
+        init {
+            this.adapter = adapter(moshi, types[0])
+        }
+
+        @Throws(IOException::class)
+        override fun fromJson(reader: JsonReader): RestResult<T>? {
+            reader.beginObject()
+
+            var result: T? = null
+            var total: Long? = null
+            var offset: Long? = null
+            var count: Long? = null
+            while (reader.hasNext()) {
+                when (reader.selectName(OPTIONS)) {
+                    0, 1 -> {
+                        // Just ignore status or success value, we are parsing 200 OK messages
+                        reader.skipValue()
+                    }
+                    2 -> {
+                        total = reader.nextLong()
+                    }
+                    3 -> {
+                        offset = reader.nextLong()
+                    }
+                    4 -> {
+                        count = reader.nextLong()
+                    }
+                    -1 -> {
+                        reader.nextName()
+                        val token = reader.peek()
+                        if (token == JsonReader.Token.BEGIN_ARRAY || token == JsonReader.Token.BEGIN_OBJECT) {
+                            result = this.adapter.fromJson(reader)
+                        } else {
+                            reader.skipValue()
+                        }
+                    }
+                }
+            }
+
+            reader.endObject()
+
+            return RestResult(result, total, offset, count)
+        }
+
+        @Throws(IOException::class)
+        override fun toJson(writer: JsonWriter, value: RestResult<T>?) {
+        }
```

add a `TODO()` call, so if for some reason we start using, will get a crash instead of silently ignore the data.